File: go/minh/minh.go
package minh

import (
	"fmt"
	"reflect"
	"strings"
	"math"
	"runtime"
	
	minnow "github.com/phil-mansfield/minnow/go"

	"unsafe"
)

const (
	Magic = 0xbaff1ed
	Version = 0
)

const (
	basicFileType    int64 = iota
	boundaryFileType
)

const (
	Int64 int64 = iota
	Int32
	Int16
	Int8
	Uint64
	Uint32
	Uint16
	Uint8
	Float64
	Float32
	Int
	Float
)

type Writer struct {
	f *minnow.Writer
	blocks int
	cols []Column
	blockSizes []int64
	buf []float32
	l, boundary float32
	cells int
}

type Column struct {
	Type int64
	Log int32
	Low, High, Dx float32
	Buffer [232]byte
}

func (c Column) String() string {
	return fmt.Sprintf("{Type: %s, Log: %t, Range: (%g %g), Dx: %g}",
		minnow.GroupNames[c.Type], c.Log != 0, c.Low, c.High, c.Dx)
}
func (c Column) GoString() string { return c.String() }

type idHeader struct {
	Magic, Version, FileType int64
}

type geometry struct {
	L, Boundary float32
	Cells int64
}

func Create(fname string) *Writer {
	if unsafe.Sizeof(Column{}) != 256 {
		panic(fmt.Sprintf("Sizeof(Column{}) = %d, not 256. Change buffer size.",
			unsafe.Sizeof(Column{})))
	}

	wr := &Writer{
		f: minnow.Create(fname),
	}
	wr.f.Header(idHeader{ Magic, Version, basicFileType })
	wr.f.Header(Boundary{ })
	return wr
}

func (minh *Writer) Header(names []string, text string, cols []Column) {
	minh.f.Header([]byte(strings.Join(names, "$")))
	minh.f.Header([]byte(text))
	minh.f.Header(cols)
	minh.cols = cols
}

func (minh *Writer) Geometry(L, boundary float32, cells int) {
	minh.l, minh.boundary, minh.cells = L, boundary, cells 
}

func (minh *Writer) Block(cols []interface{}) {
	if len(cols) != len(minh.cols) {
		panic(fmt.Sprintf("Expected %d columns, got %d.", 
			len(minh.cols), len(cols)))
	}
	for i := range cols {
		colType := minh.cols[i].Type
		if err := minnow.TypeMatch(cols[i], colType); err != nil {
			panic(fmt.Sprintf("Column %d: %s", i, err.Error()))
		}
	}
	
	N := reflect.ValueOf(cols[0]).Len()
	minh.blockSizes = append(minh.blockSizes, int64(N))
	minh.blocks++

	for i := range cols {
		if Ni := reflect.ValueOf(cols[0]).Len(); N != Ni {
			panic(fmt.Sprintf("len(cols[%d]) = %d instead of %d", Ni, i, N))
		}

		colType := minh.cols[i].Type
		switch {
		case colType >= minnow.Int64Group && colType <= minnow.Float32Group:
			minh.f.FixedSizeGroup(colType, N)
			minh.f.Data(cols[i])
		case colType == minnow.IntGroup:
			minh.f.IntGroup(N)
			minh.f.Data(cols[i])
		case colType == minnow.FloatGroup:
			lim := [2]float32{ minh.cols[i].Low, minh.cols[i].High }
			x := cols[i].([]float32)
			minh.buf = expandFloat32(minh.buf, len(x))
			for j := range x {
				minh.buf[j] = x[j]

				if minh.cols[i].Log != 0 {
					minh.buf[j] = float32(math.Log10(float64(minh.buf[j])))
				}
				if minh.buf[j] < minh.cols[i].Low {
					minh.buf[j] = minh.cols[i].Low
				}
				if minh.buf[j] >= minh.cols[i].High {
					minh.buf[j] = math.Nextafter32(
						minh.cols[i].High, float32(math.Inf(-1)),
					)
				}
			}

			minh.f.FloatGroup(N, lim, minh.cols[i].Dx)
			minh.f.Data(minh.buf)
		}
	}
}

func (minh *Writer) Close() {
	minh.f.Header(boundary{ minh.l, minh.boundary, int64(minh.cells) })
	minh.f.Header(int64(minh.blocks))
	minh.f.Header(minh.blockSizes)
	minh.f.Close()
}

func expandFloat32(buf []float32, N int) []float32 {
	if cap(buf) >= N { return buf[:N] }
	buf = buf[:cap(buf)]
	return append(buf, make([]float32, N - len(buf))...)
}

func expandInt64(buf []int64, N int) []int64 {
	if cap(buf) >= N { return buf[:N] }
	buf = buf[:cap(buf)]
	return append(buf, make([]int64, N - len(buf))...)
}

type Reader struct {
	Names []string
	Text string
	Columns []Column
	Blocks, Length int
	BlockLengths []int
	L, Boundary float32
	Cells int


	f *minnow.Reader
	fileType int64
}

func Open(fname string) *Reader {
	f := minnow.Open(fname)
	hd := &idHeader{ }
	f.Header(0, hd)

	if hd.Magic != Magic {
		panic(fmt.Sprintf("%s is not a minh file. Expected magic number " + 
			"%d, but got %d.", fname, Magic, hd.Magic))
	} else if hd.Version < Version {
		panic(fmt.Sprintf("%s written with minh version %d, but reader " + 
			"is version %d.", fname, hd.Version, Version))
	}

	byteNames := make([]byte, f.HeaderSize(1))
	byteText := make([]byte, f.HeaderSize(2))
	cols := make([]Column, f.HeaderSize(3)/int(unsafe.Sizeof(Column{})))
	geom := &geometry{ }
	i64Blocks := int64(0)
	i64BlockLengths := make([]int64, f.HeaderSize(6) / 8)
	
	f.Header(1, byteNames)
	f.Header(2, byteText)
	f.Header(3, cols)
	f.Header(4, geom)
	f.Header(5, &i64Blocks)
	f.Header(6, i64BlockLengths)
	
	minh := &Reader{
		f: f,
		fileType: hd.FileType,
		Names: strings.Split(string(byteNames), "$"),
		Text: string(byteText),
		Columns: cols,
		Blocks: int(i64Blocks),
		BlockLengths: make([]int, len(i64BlockLengths)),
		L: geom.L,
		Boundary: geom.Boundary,
		Cells: geom.Cells,
	}

	for i := 0; i < len(i64BlockLengths); i++ {
		minh.BlockLengths[i] = int(i64BlockLengths[i])
		minh.Length += int(i64BlockLengths[i])
	}

	return minh
}

func (rd *Reader) Ints(names []string) map[string][]int64 {
	out := map[string][]int64{ }
	for _, name := range names { out[name] = make([]int64, rd.Length) }

	end := 0

	for b := 0; b < rd.Blocks; b++ {
		start := end
		end = start + rd.BlockLengths[b]
		bOut := map[string][]int64{ }

		for _, name := range names { bOut[name] = out[name][start:end] }
		rd.IntBlock(b, bOut)
	}

	return out
}

func (rd *Reader) Floats(names []string) map[string][]float32 {
	out := map[string][]float32{ }
	for _, name := range names { out[name] = make([]float32, rd.Length) }

	end := 0

	for b := 0; b < rd.Blocks; b++ {
		start := end
		end = start + rd.BlockLengths[b]
		bOut := map[string][]float32{ }

		for _, name := range names { bOut[name] = out[name][start:end] }
		rd.FloatBlock(b, bOut)
	}

	return out
}

func (rd *Reader) IntBlock(b int, out map[string][]int64) {
	runtime.GC()
	for name, arr := range out {
		arr = expandInt64(arr, rd.BlockLengths[b])

		if len(arr) != rd.BlockLengths[b] {
			panic(fmt.Sprintf("Reader.BlockLengths[%d] = %d, but " + 
				"len(out[%s]) = %d", rd.BlockLengths[b], b, name, len(arr)))
		}

		var idx int
		c := findName(name, rd.Names)
		if rd.fileType == basicFileType {
			idx = c + b*len(rd.Columns)
		} else {
			idx = c*rd.Blocks + b
		}

		if err := minnow.TypeMatch(arr, rd.Columns[c].Type); err != nil {
			panic(fmt.Sprintf("Column '%s': %s", name, err.Error()))
		}

		rd.f.Data(idx, arr)

		out[name] = arr
	}
}

func (rd *Reader) FloatBlock(b int, out map[string][]float32) {
	runtime.GC()
	for name, arr := range out {
		arr = expandFloat32(arr, rd.BlockLengths[b])

		c := findName(name, rd.Names)
		var idx int
		if rd.fileType == basicFileType {
			idx = c + b*len(rd.Columns)
		} else {
			idx = c*rd.Blocks + b
		}

		if err := minnow.TypeMatch(arr, rd.Columns[c].Type); err != nil {
			panic(fmt.Sprintf("Column '%s': %s", name, err.Error()))
		}

		rd.f.Data(idx, arr)

		if rd.Columns[c].Log != 0 {
			for i := range arr {
				arr[i] = float32(math.Pow(10, float64(arr[i])))
			}
		}

		out[name] = arr
	}
}

func (rd *Reader) Close() {
	rd.f.Close()
}

func findName(name string, names []string) int {
	for i := range names {
		if name == names[i] { return i }
	}
	panic(fmt.Sprintf("Name %s not in Reader.Names = %s.", name, names))
}

File: python/minh.py
from __future__ import print_function

import minnow
import struct
import numpy as np
import gc

MAGIC = 0xbaff1ed
VERSION = 0

_basic_file_type = 0
_boundary_file_type = 1

_column_buf_size = 232
_column_type = np.dtype([
    ("type", np.int64),
    ("log", np.int32),
    ("low", np.float32),
    ("high", np.float32),
    ("dx", np.float32),
    ("buf", "S%d" % _column_buf_size)
])
assert(_column_type.itemsize == 256)

def create(fname):
    return Writer(fname)

def open(fname):
    return Reader(fname)

class Column(object):
    def __init__(self, type, log=0, low=0, high=0, dx=0):
        self.type, self.log = type, log != 0
        self.low, self.high, self.dx = low, high, dx
        
class Writer(object):
    def __init__(self, fname):
        self.f = minnow.create(fname)
        self.f.header(struct.pack("<qqq", MAGIC, VERSION, _basic_file_type))
        self.blocks = 0
        self.cols = None
        self.block_sizes = []
        self.cells, self.L, self.boundary = 0, 0, 0

    def header(self, names, text, cols):
        self.cols = cols
        self.f.header(text.encode("ascii"))
        self.f.header("$".join(names).encode("ascii"))
        
        bin_cols = np.zeros(len(cols), dtype=_column_type)
        for i in range(len(cols)):
            bin_cols["type"][i] = cols[i].type
            bin_cols["log"][i] = cols[i].log
            bin_cols["low"][i] = cols[i].low
            bin_cols["high"][i] = cols[i].high
            bin_cols["dx"][i] = cols[i].dx
        
        self.f.header(bin_cols)

    def geometry(self, L, boundary, cells):
        self.L, self.boundary, self.cells = L, boundary, cells

    def block(self, cols):
        assert(len(cols) == len(self.cols))
        for i in range(len(cols)):
            assert(minnow.type_match(self.cols[i].type, cols[i]))

        self.block_sizes.append(len(cols[0]))
        self.blocks += 1

        for i in range(len(cols)):
            assert(len(cols[i]) == len(cols[0]))
            col_type = self.cols[i].type

            if (col_type >= minnow.int64_group and
                col_type <= minnow.float32_group):
                self.f.fixed_size_group(col_type, len(cols[i]))
                self.f.data(cols[i])
            elif col_type == minnow.int_group:
                self.f.int_group(len(cols[i]))
                self.f.data(cols[i])
            elif col_type == minnow.float_group:
                lim = (self.cols[i].low, self.cols[i].high)
                buf = np.asarray(np.copy(cols[i]), dtype=np.float32)
                if self.cols[i].log: np.log10(buf, out=buf)
                buf[buf >= self.cols[i].high] = np.nextafter(
                    self.cols[i].high, np.float32(-np.inf), dtype=np.float32
                )
                buf[buf < self.cols[i].low] = self.cols[i].low
                
                self.f.float_group(len(cols[i]), lim, self.cols[i].dx)
                self.f.data(buf)
            
    def close(self):
        self.f.header(struct.pack("<ffq", self.L, self.boundary, self.cells))
        self.f.header(struct.pack("<q", self.blocks))
        self.f.header(np.array(self.block_sizes, dtype=np.int64))
        self.f.close()

class Reader(object):
    def __init__(self, fname):
        self.f = minnow.open(fname)

        magic, version, self.file_type = self.f.header(0, "qqq")
        assert(magic == MAGIC)
        assert(version == VERSION)

        self.text = self.f.header(1, "s")
        self.names = self.f.header(2, "s")
        raw_columns = self.f.header(3, _column_type)
        self.L, self.boundary, self.cells = self.f.header(4, "ffq")
        self.blocks = self.f.header(5, "q")
        self.block_lengths = self.f.header(6, np.int64)

        self.columns = [None]*len(raw_columns)
        for i in range(len(raw_columns)):
            self.columns[i] = Column(
                raw_columns["type"][i], raw_columns["log"][i], 
                raw_columns["low"][i], raw_columns["high"][i], 
                raw_columns["dx"][i]
            )
            
        self.names = self.names.split("$")

        self.length = np.sum(self.block_lengths)

    def is_boundary(self):
        return self.cells > 0

    def read(self, names):
        blocked_out = [[None]*self.blocks for _ in range(len(names))]
        for b in range(self.blocks):
            block = self.block(b, names)
            for n in range(len(names)):
                blocked_out[n][b] = block[n]

        out = [None]*len(names)
        for n in range(len(names)):
            out[n] = np.hstack(blocked_out[n])

        return out

    def block(self, b, names):
        gc.collect()
        out = [None]*len(names)

        for i in range(len(names)):
            c = self.names.index(names[i])
            assert(c >= 0)
            
            if self.file_type == _basic_file_type:
                idx = b*len(self.columns) + c
            else:
                idx = b + c*self.blocks

            out[i] = self.f.data(idx)

            if self.columns[c].log: out[i]=10**out[i]

        return out


    def close(self):
        self.f.close()

    def block_origin(self, b):
        origin =  self.cell_origin(b) - self.boundary
        origin[origin < 0] += self.L
        return origin

    def block_width(self):
        return self.cell_width() + self.boundary*2

    def cell_origin(self, b):
        ix = b % self.cells
        iy = (b // self.cells) % self.cells
        iz = b // (self.cells * self.cells)
        return np.array([ix, iy, iz]) * self.cell_width()

    def cell_width(self):
        return self.L / self.cells
